### PR TITLE
Performance improvements

### PR DESCRIPTION
**Pull request by jwidauer**
As mentioned in my previous PR, here are some performance improvements that I've made to the library.
The improvements mostly targeted the ppk functionality of rtklib, but should also improve the performance of other modes of operation.

Here's some of the improvements made:
* When tracing is disabled (`-DTRACE` NOT defined), we were still doing a ton of the string formatting computations, since it wasn't fully optimized out. I changed the implementation from a noop method, to a macro that will make sure anything that is called within the parentheses of any of the `trace*` methods is compiled out. This improved ppk performance ~30% for my test cases.
* We were previously copying the state vector and covariance matrix into temporaries, before running the `filter` method on it, to ensure that the originals wouldn't be modified if something went wrong. I changed the filter method, to just not update the input, if there was a failure to compute the kalman filter update step. This enabled me to fully remove the temporaries and gain ~20% computational performance for my test cases.
* Continuing with the theme of reducing memory overhead, I noticed that the `H` matrix used in the `filter` methods, was incredibly sparse with my data sets (~70 out of ~300000 elements were non zero). I therefore implemented a sparse matrix structure and associated methods, to reduce this overhead. This addition improved performance by another ~10% to the original in my test benchmarks.
* I noticed that the dot product method was not being auto-vectorized, contained a loop and showed when profiling the benchmarks. I therefore implemented specialized methods for two and three dimensional vectors, which improved performance another ~5% for my benchmarks.

So in total I believe these changes should improve performance quite significantly and reduce computation time for ppk by a big margin.
I made some other performance improvements, by swapping out the C implementations of some key matrix methods, with C++ implementations using the `Eigen`. If there's interest in this, I'd be more than happy to also open a PR for those improvements.

I'd be happy to answer any questions or concerns and hope these improvements help!

### CODE REVIEW

File: src/rtkpos.c [hunk 1]
```diff
@@ -1180,15 +1183,15 @@ static int test_sys(int sys, int m)
...
                  double *azel, double *freq, const int *iu, const int *ir,
-                 int ns, double *v, double *H, double *R, int *vflg)
+                 int ns, double *v, sparse_mat_t *H, double *R, int *vflg)
 {
     prcopt_t *opt=&rtk->opt;
     double bl,dr[3],posu[3],posr[3],didxi=0.0,didxj=0.0,*im,icb,threshadj;
-    double *tropr,*tropu,*dtdxr,*dtdxu,*Ri,*Rj,freqi,freqj,*Hi=NULL,df;
-    int i,j,k,m,f,nv=0,nb[NFREQ*NSYS*2+2]={0},b=0,sysi,sysj,nf=NF(opt);
+    double *tropr,*tropu,*dtdxr,*dtdxu,*Ri,*Rj,freqi,freqj,df;
+    int i,j,k,m,f,nv=0,nb[NFREQ*4*2+2]={0},b=0,sysi,sysj,nf=NF(opt);
```
Comment (ourairquality) [inline src/rtkpos.c:1194]: Could this 'NFREQ*4' be retained as 'NFREQ*NSYS'?

Comment (jwidauer) [reply src/rtkpos.c:1194]: Oh yes, of course! That was an oversight on my part when pulling these changes over from an out of date branch.

**Comment by rtklibexplorer** [comment]
I tested this PR using a PPK solution for [this dataset](https://rtkexplorer.com/download/u-blox-f9p-kinematic-ppp-data-set-12-24-20/) with rover.obs and tmg23590.20o using the included ppk config file, changed only from a combined solution to a forward-only solution.  I compared results from the code with and without these changes.  For the most part the solutions were nearly identical, but they diverged in two places and the fix rate dropped from 96.0% to 90.3%.  Could you duplicate the experiment, and see if you see the same thing?  I am concerned about pulling this change until this is resolved.

I tested with the RTKPOST app, built with the Embarcadero compiler with -DTRACE enabled.  In this configuration, I saw only a 13% reduction in processing time.  I suspect some of the improvements will be compiler dependent.

**Comment by jwidauer** [comment]
Thanks for testing it out!

I went and tested it on that dataset as well, but used the `rnx2rtkp` console application, since I can't currently build the RTKPOST application, due to not having Qt.
I used this command: `./rnx2rtkp -k tmp/ppk.conf -o out.pos tmp/rover.obs tmp/tmg23590.20o tmp/rover.nav`
In addition to your change to only use `pos1-soltype =forward`, I had to also change the `file-satantfile =...` to the one with the same name in the `RTKLIB/data/ant/` directory, to make it work.
I then checked the output file using a very simple python application that just computes the fix rate by using the nr. of `1` characters in the 6th column (Q) in the output file over the total number of non header lines in the file.

Using this I found a fix rate of 95.91% using the code from this PR and a fix rate of 95.53% using the code from the `demo5` branch. So I, unfortunately, couldn't reproduce your findings. Do you see anything obvious that I did wrong while testing?

I then also timed the different versions of the previous command, each ran 10 times and these are the average run times recorded by the `time` command:
* without improvements, with `-DTRACE`: 5.474s (100%)
* with improvements, with `-DTRACE`: 4.797s (~87.6%)
* with improvements, without `-DTRACE`: 2.688s (49.1%)

So, as you can see, yes, it seems like with `-DTRACE` enabled, the improvement is only ~13% on this specific dataset. I suspect that it uses slightly different code paths than the sample I used for my benchmarks, which is where the difference comes from.
It also shows that having `-DTRACE` enabled, is a MAJOR performance bottleneck, which might be best recommended to be disabled, or potentially refactored further.

I also rebased this branch on top of the latest changes on the `demo5` branch, so it should be up to date again.

**Comment by jwidauer** [comment]
After a bit of thought, I came up with a solution that should also improve performance quite drastically with `-DTRACE` enabled.
Please have a look and let me know what you think! I'm happy to also remove it, if deemed ugly or else!

The option with `-DTRACE` disabled, is still a bit faster, but the difference is significantly less than before!

File: src/rtkcmn.c [hunk 2]
```diff
@@ -1152,15 +1173,15 @@ extern void matmul(const char *tr, int n, int k, int m, double alpha,
...
+    for (j=0;j<k;j++) for (i=0;i<n;i++) {
         d=0.0;
         switch (f) {
-            case 1: for (x=0;x<m;x++) d+=A[i+x*n]*B[x+j*m]; break;
+            case 0: for (x=0;x<m;x++) d+=A[i+x*n]*B[x+j*m]; break;
+            case 1: for (x=0;x<m;x++) d+=A[x+i*m]*B[x+j*m]; break;
             case 2: for (x=0;x<m;x++) d+=A[i+x*n]*B[j+x*k]; break;
-            case 3: for (x=0;x<m;x++) d+=A[x+i*m]*B[x+j*m]; break;
-            case 4: for (x=0;x<m;x++) d+=A[x+i*m]*B[j+x*k]; break;
+            case 3: for (x=0;x<m;x++) d+=A[x+i*m]*B[j+x*k]; break;
```
Comment (ourairquality) [inline src/rtkcmn.c:1184]: fwiw the changes to matmul() appear to be self contained, could be split out into a separate PR if that helped. The main change appears to be reordering the loop memory access, perhaps to optimize for some of the cases.

File: src/rtkcmn.c [hunk 3]
```diff
@@ -1020,6 +1022,24 @@ extern double *eye(int n)
     if ((p=zeros(n,n))) for (i=0;i<n;i++) p[i+i*n]=1.0;
     return p;
 }
+
+/* dot product -----------------------------------------------------------------
+ * inner product of vectors of size 2
+ * args   : double *a,*b     I   vectors a and b
+ * return : a'*b
+ *-----------------------------------------------------------------------------*/
+extern double dot2(const double* a, const double* b) { return a[0] * b[0] + a[1] * b[1]; }
```
Comment (ourairquality) [inline src/rtkcmn.c:1031]: It appears that the dot product patch could be split out into a separate PR, perhaps just a case of creating a new branch off demo5 and a git cherry-pick.

**Comment by ourairquality** [comment]
The trace improvements look useful alone and I suggest splitting them out into a separate PR so they can be evaluated separately from the sparse matrix improvements.

**Comment by rtklibexplorer** [comment]
Yes, I think that makes sense.  Meanwhile, I will take another look at this code to see if I can better understand the differences in our results.

File: src/rtkcmn.c (hunk 2 above)
Comment (jwidauer) [reply src/rtkcmn.c:1184]: Yep, that's exactly right. This pretty much just changes the memory access order. Which then provided around ~5% performance improvements in my test cases (using it for ppk with both forward and backward passes).
I opened #197 for these and the changes that remove the unnecessary temporary used in `rtkpos.c`. Have a look and let me know what you think!

File: src/rtkcmn.c (hunk 3 above)
Comment (jwidauer) [reply src/rtkcmn.c:1031]: Done! That's #196 now!

**Comment by ourairquality** [comment]
With this the sparse matrix patch alone, I see around a 1% increase in run time and some changes in the results which I assume is unexpected.

**Comment by jwidauer** [comment]
So I ran a couple more tests with different compilers. My test setup is the following:
I run `rnx2rtkp` with the following command: `./rnx2rtkp -k tmp/ppk.conf -o out/out.pos tmp/rover.obs tmp/tmg23590.20o tmp/rover.nav`
The dataset I use is [the one provided above](https://rtkexplorer.com/download/u-blox-f9p-kinematic-ppp-data-set-12-24-20/), with only the `file-satantfile` option in the `ppk.conf` changed to `RTKLIB/data/ant/igs14.atx` (so I'm running the combined ppk).
These are run 10 times and then I compute the average and standard deviation. These are my results when comparing this PR (now with only the sparse matrix change) to the base (demo5):

| Compiler | Base avg. (s) | Base std. (s) | Sparse avg. (s) | Sparse std. (s) | Improvement |
| - | - | - | - | - | - |
| GCC 11.4.0 | 6.19 | 0.02 | 6.08 | 0.01 | 1.73% |
| GCC 12.3.0 | 5.98 | 0.01 | 5.85 | 0.01 | 2.17% |
| Clang 14.0.0 | 5.7 | 0.02 | 5.36 | 0.01 | 5.89% |
| Clang 15.0.7 | 5.54 | 0.02 | 5.38 | 0.01 | 2.83% |

I believe the reason why the impact is so low is that the computation time is still dominated by the copy of the big state covariance matrix that I addressed in PR #197, as well as the hand rolled implementation of the matrix multiplication.

I do also see the difference in output, which surprises me, since I was very careful about checking for any output changes, while developing this. I will have to look into this some more to see whether I made a mistake when migrating these changes from our internal implementation. If you have any idea as to what could cause these changes, I'd be very happy to hear!

**Comment by ourairquality** [comment]
Noticed one issue in rtkpoc.c ddres(), the previous code was zeroing as it goes, and that still appears necessary. See the attached patch to show the issue and correct it, and with this I see no difference in results.
[patch-ddres.txt](https://github.com/rtklibexplorer/RTKLIB/files/14531293/patch-ddres.txt)

Btw do other people see seleph() dominating performance now, it's around 30% on one long run test, and that is probably swamping out improvements in that test case.

**Comment by jwidauer** [comment]
I thought I had added that back in the last change, but it seems I had mixed up the rows and columns. I changed it now and also don't see a change in output anymore! Thanks for the heads up!

I profiled the application with the sparse matrix change in this PR and the setup I described above as well, but I don't see `seleph()` dominating at all. Neither on gcc 11.4.0 nor gcc 12.3.0.
For me the dominating factor is still the memory copy addressed in PR #197, followed by `matmul()` and `matinv()`.
Could you give me some additional information about your setup and test cases, so I might be able to try and reproduce your findings?

File: src/rtkpos.c [hunk 4]
```diff
@@ -1593,17 +1600,18 @@ static void holdamb(rtk_t *rtk, const double *xa)
     for (i=0;i<nv;i++) R[i+i*nv]=rtk->opt.varholdamb;
         
     /* update states with constraints */
+    sparse_mat_compress(H);
```
Comment (ourairquality) [inline src/rtkpos.c:1603]: The only consumer of the sparse matrix appears to be filter(), and here the sparse matrix is freed after this call, so it appears that the call to sparse_mat_compress() could be avoided here.

File: src/rtkpos.c [hunk 5]
```diff
@@ -2024,6 +2033,7 @@ static int relpos(rtk_t *rtk, const obsd_t *obs, int nu, int nr,
                 xp=x+K*v
                 Pp=(I-K*H')*P                  */
         trace(3,"before filter x=");tracemat(3,rtk->x,1,9,13,6);
+        sparse_mat_compress(H);
```
Comment (ourairquality) [inline src/rtkpos.c:2036]: Although this is in a loop, there is typically only one iteration. Also the sparse matrix does not appear to be used after this loop, so sparse_mat_free(H) could be mode up to just after this loop. The only consumer of the sparse matrix appears to be filter(). So it might be okay to omit the call to sparse_mat_compress() here too.

**Comment by ourairquality** [comment]
To see even a slight performance improvement, it was necessary to add an efficient sparse matrix column clearing function and use that in ddres() to replace the zero filling, and to remove the sparse matrix compact calls, and to create and free the sparse matrix in relpos() on each loop calling ddres().

Also tried adding an efficient function to clear the sparse array, keeping the internal arrays, but it was little different to freeing and creating a new sparse array.

The sparse matrix created in holdamb() has only 2 elements per column, and the matrix used in ddres() appears to have at most 5 elements per column. Also there are no writes to existing elements, no duplicates. Perhaps it would be more efficient to pre-allocate fixed sized columns, to save all the adjustments to the columns start indexes, and the data moves. Perhaps also to not bother sorting the elements in a column, to not bother looking for duplicates, to not bother removing elements written with value zero, and see if the consumer filter() can just iterate over the index and value.

**Comment by rtklibexplorer** [comment]
Closing this PR as it has been broken into pieces and I believe those pieces have been addressed in subsequent PRs.
